### DISCUSSION TITLE
Tips for code formatting on VSCode?

### DISCUSSION THREAD

**Discussion by javi7** [teal-language/tl#654] [Q&A] [answered]
I just started playing around with teal this week and I'm loving it. I'm planning to use it to make games in LOVE 2D. 

I want to set up code formatting in VSCode so that tabs and whitespace rules can be automatically fixed on save. Has anyone had any success doing this? Or could point me to some resources to get started?

**Comment by pdesaulniers** [comment, marked as answer]
Hello!

The VSCode extension does not currently support this. It does support a basic form of auto-indenting (through `increaseIndentPattern` and `decreaseIndentPattern` in `language-configuration.json`) but it is very unreliable. You can try it out manually with `Ctrl+Shift+P > Reindent Lines`.

I think the best way to tackle this would be to implement some sort of formatting tool, as proposed in https://github.com/teal-language/tl/issues/84. Then, the next step would be to integrate this tool into the VSCode extension.

I haven't looked into it, but I assume you would need to parse the document using Teal's API, then convert the AST back into code, based on some formatting rules.
